File: src/test/java/info/skyblond/archivedag/service/impl/CertServiceTest.java
package info.skyblond.archivedag.service.impl;

import info.skyblond.archivedag.model.entity.CertEntity;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.AuthenticationException;
import org.springframework.test.context.ActiveProfiles;

import java.security.cert.X509Certificate;

import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
@ActiveProfiles("test")
class CertServiceTest {

    @Autowired
    CertService certService;

    @Test
    void testVerifyCertStatus() throws Exception {
        X509Certificate others = this.certService.signCert("test_user_404").getCertificate();
        X509Certificate locked = this.certService.signCert("test_user").getCertificate();
        assertEquals(CertEntity.Status.LOCKED, this.certService.queryCert(getSerialNum(locked)).getStatus());
        X509Certificate revoked = this.certService.signCert("test_user").getCertificate();
        this.certService.changeCertStatus(getSerialNum(revoked), CertEntity.Status.REVOKED);
        assertEquals(CertEntity.Status.REVOKED, this.certService.queryCert(getSerialNum(revoked)).getStatus());
        X509Certificate disabled = this.certService.signCert("test_user").getCertificate();
        this.certService.changeCertStatus(getSerialNum(disabled), CertEntity.Status.DISABLED);
        assertEquals(CertEntity.Status.DISABLED, this.certService.queryCert(getSerialNum(disabled)).getStatus());
        X509Certificate enabled = this.certService.signCert("test_user").getCertificate();
        this.certService.changeCertStatus(getSerialNum(enabled), CertEntity.Status.ENABLED);
        assertEquals(CertEntity.Status.ENABLED, this.certService.queryCert(getSerialNum(enabled)).getStatus());

        assertThrows(BadCredentialsException.class,
                () -> this.certService.verifyCertStatus(getSerialNum(others), "test_user"));
        assertThrows(AuthenticationException.class,
                () -> this.certService.verifyCertStatus(getSerialNum(locked), "test_user"));
        assertThrows(AuthenticationException.class,
                () -> this.certService.verifyCertStatus(getSerialNum(revoked), "test_user"));
        assertThrows(AuthenticationException.class,
                () -> this.certService.verifyCertStatus(getSerialNum(disabled), "test_user"));
        assertDoesNotThrow(() -> this.certService.verifyCertStatus(getSerialNum(enabled), "test_user"));
    }

    private String getSerialNum(X509Certificate certificate) {
        return certificate.getSerialNumber().toString(16);
    }
}

File: backend/src/test/kotlin/info/skyblond/archivedag/arstue/CertServiceTest.kt
package info.skyblond.archivedag.arstue

import info.skyblond.archivedag.arstue.entity.CertEntity
import info.skyblond.archivedag.arstue.service.CertSigningConfigService
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.core.AuthenticationException
import org.springframework.test.context.ActiveProfiles
import java.security.KeyStore
import java.security.cert.X509Certificate
import java.security.interfaces.RSAPrivateKey
import java.security.interfaces.RSAPublicKey
import javax.net.ssl.TrustManagerFactory
import javax.net.ssl.X509TrustManager

@SpringBootTest
@ActiveProfiles("test")
internal class CertServiceTest {

    @Autowired
    lateinit var certService: CertService

    @Autowired
    lateinit var certSigningInfo: CertSigningConfigService

    @Test
    fun testSignedCert() {
        val (_, cert, privateKey) = certService.signCert("test_user")

        // make sure private key matches the cert
        val rsaPrivateKey = privateKey as RSAPrivateKey
        val rsaPublicKey = cert.publicKey as RSAPublicKey
        assertEquals(rsaPrivateKey.modulus, rsaPublicKey.modulus)

        // then check the cert matches the CA
        val anchors = KeyStore.getInstance(KeyStore.getDefaultType())
        anchors.load(null, null)
        anchors.setCertificateEntry("cert", this.certSigningInfo.caCert)
        // set up our CA as trust anchors and init trust manager
        val trustManagerFactory = TrustManagerFactory.getInstance(
            TrustManagerFactory.getDefaultAlgorithm()
        )
        trustManagerFactory.init(anchors)
        for (trustManager in trustManagerFactory.trustManagers) {
            if (trustManager is X509TrustManager) {
                Assertions.assertDoesNotThrow {
                    trustManager.checkServerTrusted(arrayOf(cert), "RSA")
                }
            }
        }
    }

    private fun getSerialNum(certificate: X509Certificate): String {
        return certificate.serialNumber.toString(16)
    }

    @Test
    fun testVerifyCertStatus() {
        // test different status
        val disabled = certService.signCert("test_user").certificate
        certService.changeCertStatus(getSerialNum(disabled), CertEntity.Status.DISABLED)
        assertEquals(CertEntity.Status.DISABLED, certService.queryCert(getSerialNum(disabled)).status)
        val enabled = certService.signCert("test_user").certificate
        certService.changeCertStatus(getSerialNum(enabled), CertEntity.Status.ENABLED)
        assertEquals(CertEntity.Status.ENABLED, certService.queryCert(getSerialNum(enabled)).status)

        assertThrows(AuthenticationException::class.java) {
            certService.verifyCertStatus(getSerialNum(disabled), "test_user")
        }

        // test others cert
        val others = certService.signCert("test_user_404").certificate
        assertThrows(BadCredentialsException::class.java) {
            certService.verifyCertStatus(getSerialNum(others), "test_user")
        }

        // test normal cert
        Assertions.assertDoesNotThrow { certService.verifyCertStatus(getSerialNum(enabled), "test_user") }
    }

}
